{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# Title: Locating the Ori of Bacteria \n",
    "\n",
    "# Author: Surur Khan \n",
    "\n"
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Why Try To Locate the Ori?\n",
    "\n",
    "1. Gene Therapy - intentionally infect a patient who lacks a crucial gene with a viral vector containing an artificial gene that encodes a therapeutic protein. Once inside the cell, the vector replicates and eventually produces many copies of the therapeutic protein, which in turn treats the patient’s disease. To ensure that the vector actually replicates inside the cell, biologists must know where ori is in the vector’s genome and ensure that the genetic manipulations that they perform do not affect it(This usage is also referred to as Genetic Engineering). The scripts were tested against an E.Coli Genome dataset from the NCBI and yielded significant results for finding it's ori by analyzing the GC content changes and implications of a potential 9mer DnaA box due to a conservative sequence found in a relatively short segment of the Genome.\n",
    "\n",
    "2. Ensure Genetic Manipulations do not affect overall cell functions \n",
    "\n",
    "# What I Test on\n",
    "\n",
    "The package Bioezy (v 0.0.4) contains several scripts for analyzing genomic data, mainly to locate replication origins. \n",
    "\n",
    "## For this analysis, the bacterium **E. Coli (Vibrio Cholerae)** is used. For Credibility, the nucleotide sequence appearing in the ori of E. Coli was predetermined as follows:\n",
    "\n",
    "atcaatgatcaacgtaagcttctaagcatgatcaaggtgctcacacagtttatccacaac\n",
    "ctgagtggatgacatcaagataggtcgttgtatctccttcctctcgtactctcatgacca\n",
    "cggaaagatgatcaagagaggatgatttcttggccatatcgcaatgaatacttgtgactt\n",
    "gtgcttccaattgacatcttcagcgccatattgcgctggccaaggtgacggagcgggatt\n",
    "acgaaagcatgatcatggctgttgttctgtttatcttgttttgactgagacttgttagga\n",
    "tagacggtttttcatcactgactagccaaagccttactctgcctgacatcgaccgtaaat\n",
    "tgataatgaatttacatgcttccgcgacgatttacctcttgatcatcgatccgattgaag\n",
    "atcttcaattgttaattctcttgcctcgactcatagccatgatgagctcttgatcatgtt\n",
    "tccttaaccctctattttttacggaagaatgatcaagctgctgctcttgatcatcgtttc\n",
    "\n",
    "hereinafter referred to as **vibrio_cholerae_ori.txt**\n",
    "\n",
    "The actual Genome spans 1,108,250 nucleotides long, which can be found at:\n",
    "\n",
    "[E.Coli_Genome](https://bioinformaticsalgorithms.com/data/realdatasets/Replication/Vibrio_cholerae.txt)\n",
    "\n",
    "I will not be analyzing this, rather, I will be using a segment for the sake of runtime, denoted **\"vibrio_cholerae_genome_short\"**\n",
    "\n",
    "# What I Look for\n",
    "\n",
    "There is no shortage of agreement that DNA replication occurs in all cells as it is the nature of living organisms. This however begs the question, how does the cell know where to begin replication in this short region within the otherwise huge genome? It is known that a protein that initiates replication, **DnaA** by binding to a short segment in the ori, **DnaA Box**. My goal to locate these DnaA Boxes is based on a theorum by William Legrand in Edgar Allan Poe's story \"The Gold-Bug\":\n",
    "\n",
    "_Assuming DNA is a language of its own, locate frequent words within the ori because for various biological processes, certain nucleotide strings often appear surprisingly often in small regions of the genome_\n",
    "This is often because certain proteins can only bind to DNA if a specific string of nucleotides is present, and if there are more occurrences of the string, then it is more likely that binding will successfully occur. (It is also less likely that a mutation will disrupt the binding process.)\n",
    "\n",
    "## Hence, I will extensively analyze the E. Coli Genome to look for its Ori by locating conservative sequences in a relatively short segment of the whole genome"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "source": [
    "from bioezy import bzy\n",
    "import os"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "source": [
    "print(os.getcwd())"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "/home/sururkhan/Python Code(wsl)/bioezypkg/Vibrio_Cholerae_Analysis\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "source": [
    "#Important note: frequent words is not usually used for finding frequent words because it has a runtime of O(n^2);  (|Text| − k + 1) · (|Text| − k + 1) · k == |Text|^2 · k\n",
    "#On specific ori region\n",
    "with open(\"vibrio_cholerae_ori.txt\",\"r\") as input_file:\n",
    "    vc_ori = ''.join(line.rstrip() for line in input_file).lower() #manipulation to make it 1 consecutive string\n",
    "    input_file.close()\n",
    "print(vc_ori)\n",
    "    "
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "atcaatgatcaacgtaagcttctaagcatgatcaaggtgctcacacagtttatccacaacctgagtggatgacatcaagataggtcgttgtatctccttcctctcgtactctcatgaccacggaaagatgatcaagagaggatgatttcttggccatatcgcaatgaatacttgtgacttgtgcttccaattgacatcttcagcgccatattgcgctggccaaggtgacggagcgggattacgaaagcatgatcatggctgttgttctgtttatcttgttttgactgagacttgttaggatagacggtttttcatcactgactagccaaagccttactctgcctgacatcgaccgtaaattgataatgaatttacatgcttccgcgacgatttacctcttgatcatcgatccgattgaagatcttcaattgttaattctcttgcctcgactcatagccatgatgagctcttgatcatgtttccttaaccctctattttttacggaagaatgatcaagctgctgctcttgatcatcgtttc\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "source": [
    "FreqWords = bzy.frequent_words(vc_ori,9) \n",
    "for kmer in FreqWords:\n",
    "    print(f' 9mer: {kmer} appeared {bzy.pattern_count(vc_ori,kmer)} times')\n",
    "\n"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      " 9mer: atgatcaag appeared 3 times\n",
      " 9mer: ctcttgatc appeared 3 times\n",
      " 9mer: tcttgatca appeared 3 times\n",
      " 9mer: cttgatcat appeared 3 times\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Significance\n",
    "\n",
    "We highlight a most frequent 9-mer instead of using some other value of k because experiments have revealed that bacterial DnaA boxes are usually nine nucleotides long. The probability that there exists a 9-mer appearing three or more times in a randomly generated DNA string of length 500 is approximately 1/1300.\n",
    "\n",
    "## Important observation: atgatcaag and cttgatcat are reverse complements of each other, hence **now we must account for the reverse compliment of each frequent 9mer appearing as well!**. We can strongly deduce that DnaA does not need to bind to specifically bind to either the 5'-> 3' forward(lagging) strand nor the 3' -> 5' reverse(leading) strand!\n",
    "\n",
    "Before concluding that we have found the DnaA box of Vibrio cholerae, we'll check if there are other short regions in the Vibrio cholerae genome exhibiting multiple occurrences of atgatcaag or cttgatcat. Maybe these strings occur as repeats throughout the entire Vibrio cholerae genome, rather than just in the ori region. We will use the pattern matching function to determine this.\n"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "source": [
    "print(bzy.pattern_matching(\"atgatcaag\",vc_ori)) #Locations of atgatcaag in ori\n",
    "print(bzy.pattern_matching(\"cttgatcat\",vc_ori))"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "[27, 127, 508]\n",
      "[397, 468, 525]\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "source": [
    "\"\"\"\n",
    "TESTING ON SHORT GENOME FOR ATGATCAAG\n",
    "\"\"\"\n",
    "\n",
    "with open(\"vibrio_cholerae_genome_short.txt\",\"r\") as input_file:\n",
    "    vc_genome_short = ''.join(line.rstrip() for line in input_file).lower()\n",
    "    input_file.close()\n",
    "\n",
    "#print(vc_genome)\n",
    "print(bzy.pattern_matching(\"atgatcaag\",vc_genome_short))"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "[116556, 149355, 151913, 152013, 152394, 186189, 194276, 200076, 224527, 307692, 479770, 610980, 653338, 679985, 768828, 878903, 985368]\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Significance\n",
    "\n",
    "## After solving the Pattern Matching Problem, we discover that ATGATCAAG appears 17 times in the starting positions labelled above. With the exception of the three occurrences of ATGATCAAG in ori at starting positions 151913, 152013, and 152394, no other instances of ATGATCAAG form clumps, i.e., appear close to each other in a small region of the genome.\n",
    "\n",
    "However, a definite conclusion cannot be drawn without checking if it even appears in known ori regions from other bacteria. The case may be that the clumps of my 9mers are statistically insignificant and have nothing to do with replication. \n",
    "\n",
    "## Now, the bacterium **Thermotoga petrophila**, an extremophile will be examined to locate it's ori. A potential discovery may be that different bacterium have different oris."
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "source": [
    "with open(\"thermotoga_petrophila.txt\",\"r\") as input_file:\n",
    "    tp_ori = ''.join(line.rstrip() for line in input_file).lower()\n",
    "    input_file.close()\n",
    "\n",
    "FreqWords = bzy.frequent_words(tp_ori,9) \n",
    "for kmer in FreqWords:\n",
    "    print(f' 9mer: {kmer} appeared {bzy.pattern_count(tp_ori,kmer)} times')\n",
    "\n",
    "print(bzy.pattern_matching(\"acctaccac\",tp_ori))\n",
    "print(\"Other frequent 9mers that appear more than 2 but less than 5 times: AACCTACCA 3 AAACCTACC 3 ACCTACCAC 5 CCTACCACC 3 GGTAGGTTT 3 TGGTAGGTT 3\")"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      " 9mer: acctaccac appeared 5 times\n",
      "[184, 379, 390, 401, 479]\n",
      "Other frequent 9mers that appear more than 2 but less than 5 times: AACCTACCA 3 AAACCTACC 3 ACCTACCAC 5 CCTACCACC 3 GGTAGGTTT 3 TGGTAGGTT 3\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## It is worth noting that AACCTACCA and TGGTAGGTT are reverse complements, as are AAACCTACC and GGTAGGTTT \n",
    "\n",
    "# The Caveat\n",
    "\n",
    " Searching for “clumps” of either ATGATCAAG (reverse compliment -CTTGATCAT) or CCTACCACC(reverse compliment - GGTGGTAGG) is unlikely to help, since this new genome may use a completely different DnaA Box.\n",
    "\n",
    " Instead of finding clumps of a specific k-mer, try to find every k-mer that forms a clump in the genome, hoping they shed light on the location of ori.\n",
    "\n",
    " # The Clump Finding Problem \n",
    "\n",
    " Slide a window of fixed length L along the genome, looking for a region where a k-mer appears several times in short succession. The parameter value L = 500 reflects the typical length of ori in bacterial genomes. Define a k-mer as a \"clump\" if it appears many times within a short interval of the genome. More formally, given integers L and t, a k-mer Pattern forms an (L, t)-clump inside a (longer) string Genome if there is an interval of Genome of length L in which this k-mer appears at least t times. (This definition assumes that the k-mer completely fits within the interval. This also does not take reverse complements into account yet.)\n",
    "\n",
    " eg. TGCA forms a (25,3)-clump in the following Genome:\n",
    "gatcagcataagggtccC**TGCA**A**TGCA**TGACAAGCC**TGCA**GTtgttttac\n",
    "\n",
    "## From our previous examples of ori regions, ATGATCAAG forms a (500,3)-clump in the Vibrio cholerae genome, and CCTACCACC forms a (500,3)-clump in the Thermotoga petrophila genome."
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "source": [
    "clumps = bzy.find_clumps(vc_genome_short,9,500,3)\n",
    "print(len(clumps))\n",
    "\n",
    "#Refer to scan_clumps_eff_algo for the true solution (sourced online)"
   ],
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "acaatgagg ttcgagctc tcgagctct aaccggctg tgcgcatac gcgcatacg gccatccga cagcgtcta gtctattca ggctatgca ctatgcagg caggctact tggttcgta gtaagaact aagaacttt gaactttag gccttacct gctttagtc ctttagtcg tttagtcgt tagtcgtgg cgatctggg gtgtggctc tggctctcg cgcgtatgg gtatggtct gtctgtcta ttctaaccc aacccgcgc ccgcgctac tctgagtgc caggtctac ggtctactc gtctactcc ctactcctt cctttcggc gcactagtg gccatgggc caacgggca acgggcagc agcattact ctgttctta cttaaacgg ctggttcca ggttccaag ttccaaggc tgcttgtgg cttgtggcc ggccgtact tggtgcact ggtgcactg gtgcactgg ggtcacgca agagcgtgg gagcgtggt gtttcggtc tcggtctgg tggaacgtc\n",
      "58\n"
     ]
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Using the scan_clumps program with **vibrio_cholerae_genome_full**, more than 1600 (500,3)-clumped 9mers were found\n",
    "The result is futile in the search for DnaA Boxes, as the algorithm will locate ALL frequent kmers in the region, which may occur purely by random chance. A new method must be tested."
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "# The Analysis of the Cytosine Content and it's relevance\n",
    "\n",
    "DNA replication is an asymmetric process, the faith of the forward (leading) and reverse (lagging) half strands are different. This is because DNA Polymerase, the enzyme responsible for creating the complementary base sequence against each strand can only run in the 3' -> 5' direction. The Reverse half strand indeed runs in this direction and can hence be replicated continuously, however the same cannot be said for the forward half strand, which runs in the 5' -> 3' direction. \n",
    "\n",
    "For replication to occur on a forward half strand, the replication fork has to open by about 2000 nucleotides, allowing DNA Polymerase to replicate the strand backwards towards Ori. When this fork closes, replication terminates until it reopens. Due to the inconsistent replication, _okazaki fragments_ are generated, which are partially replicated strands of nucleotides. As a result, many primers are required to fully replicate a forward strand and DNA ligase must be used to join the strand together. The built strand is made of introns and exons, where the former along with the primers are removed by some molecular splicing process. \n",
    "\n",
    "The significance of this stems from the fact that because the forward half strand has to wait for its fork to open for replication to occur, it spends most of its life **single stranded**. Naturally, due to the absence of the strongly covalent G-C bonds that would form in a complete double helix, the nucleotide bases are exposed to much higher mutation potentials. Particularly, **Cytosine has a high likeliness to mutate into Thymine** through **deamination**. As a result, **the forward half strand has a much lower Cytosine content than the reverse half strand**. Considering the ori is made up half by the forward and the other half by the reverse, I anticipate that **half of the genome will have a decreasing Cytosine content, indicating that the forward half strand is being observed**. \n",
    "\n",
    "# Overall: \n",
    "\n",
    "#G-#C decreasing = reverse half strand (high C low G)\n",
    "\n",
    "#G-#C increasing = forward half strand (low C high G)\n",
    "\n",
    "## The analysis is now directed towards the idea that **Ori occurs where the reverse half strand transitions into the forward half strand, hence it will be located where an increasing Cytosine content begins to exponentially decrease**\n",
    "\n",
    "# Generating a Skew Diagram\n",
    "\n",
    " Compute Skewi+1(Genome) from Skewi(Genome) according to the nucleotide in position i of Genome. If this nucleotide is G, then Skewi+1(Genome) = Skewi(Genome) + 1; if this nucleotide is C, then Skewi+1(Genome)= Skewi(Genome) – 1; otherwise, Skewi+1(Genome) = Skewi(Genome).\n",
    "\n",
    " eg. CATGGGCATCGGCCATACGCC has skew diagram  0 -1 -1 -1 0 1 2 1 1 1 0 1 2 1 0 0 0 0 -1 0 -1 -2\n",
    "\n",
    " ## hence, ori will be found where the skew attains a **minimum**\n"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "source": [
    "with open(\"Ecoli_genome.txt\",\"r\") as input_file:\n",
    "    ecoli_genome = ''.join(line.rstrip() for line in input_file).lower() #manipulation to make it 1 consecutive string\n",
    "    input_file.close()\n"
   ],
   "outputs": [],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "source": [
    "# Running bzy.minimum_skew(vc_genome_short) is futile, because the algorithm is not robust enough to work with the million length sequence. It will continuously converge to a local optimum, which is not ideal\n",
    "#  For demonstration purposes, a sample input will be used\n",
    "bzy.minimum_skew(\"TAAAGACTGCCGAGAGGCCAACACGAGTGCTAGAACGAGGGGCGTAAACGCGGGTCCGAT\")\n",
    "\n",
    "#Using a more robust analysis, the ori of Vibrio Cholerae is found at approximately position 3923620 of the genome, representing the global optimum."
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "[11, 24]"
      ]
     },
     "metadata": {},
     "execution_count": 42
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    " ## Now, the objective is looking for a hidden message representing a potential DnaA box near this location. Solving the Frequent Words Problem in a window of length 500 starting at position **3923620** reveals no 9-mers (along with their reverse complements) that appear three or more times. Even if we have located ori in E. coli, it appears that the location of the DnaA Box is unanimous. \n",
    "\n",
    " ## However, accounting for mismatches, it is seen that in addition to the three occurrences of ATGATCAAG and three occurrences of its reverse complement CTTGATCAT, the Vibrio cholerae ori contains additional occurrences of ATGATCAAC and CATGATCAT, which differ from ATGATCAAG and CTTGATCAT in only a single nucleotide polymorphism. This is referred to as \"Hamming Distance\". \n",
    "\n",
    " From this, I developed the function approx_pattern_match to return every location of the occurrence of a pattern within the genome with a particular amount of mismatches. Furthermore, the function approx_pattern_count  returns the number of occurrences of a pattern in the specified genome with a particular amount of mismatches\n",
    "\n",
    " **For runtime sake, this portion will be ran on the already identified ori region of e.coli**"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "source": [
    " bzy.approx_pattern_match(vc_genome_short,\"atgatcaag\",2) \n",
    "#Note they're 2581 occurrences"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "[14,\n",
       " 180,\n",
       " 702,\n",
       " 1381,\n",
       " 1579,\n",
       " 1715,\n",
       " 1725,\n",
       " 1746,\n",
       " 1884,\n",
       " 2253,\n",
       " 2396,\n",
       " 2408,\n",
       " 2860,\n",
       " 2964,\n",
       " 3989,\n",
       " 4190,\n",
       " 4210,\n",
       " 4555,\n",
       " 4570,\n",
       " 4762,\n",
       " 7668,\n",
       " 7882,\n",
       " 8461,\n",
       " 8677,\n",
       " 8833,\n",
       " 8857,\n",
       " 9225,\n",
       " 9491,\n",
       " 10422,\n",
       " 11315,\n",
       " 11409,\n",
       " 11711,\n",
       " 11744,\n",
       " 11867,\n",
       " 12190,\n",
       " 12343,\n",
       " 13573,\n",
       " 13579,\n",
       " 14173,\n",
       " 14239,\n",
       " 15079,\n",
       " 15406,\n",
       " 15493,\n",
       " 15739,\n",
       " 17139,\n",
       " 17161,\n",
       " 18168,\n",
       " 19038,\n",
       " 19683,\n",
       " 19790,\n",
       " 20066,\n",
       " 20429,\n",
       " 20753,\n",
       " 20872,\n",
       " 21007,\n",
       " 21235,\n",
       " 21685,\n",
       " 21947,\n",
       " 22138,\n",
       " 22283,\n",
       " 22429,\n",
       " 22887,\n",
       " 23139,\n",
       " 23254,\n",
       " 23536,\n",
       " 24263,\n",
       " 24275,\n",
       " 24452,\n",
       " 24836,\n",
       " 24920,\n",
       " 25038,\n",
       " 25566,\n",
       " 25893,\n",
       " 25928,\n",
       " 26118,\n",
       " 26375,\n",
       " 26632,\n",
       " 26844,\n",
       " 27615,\n",
       " 27723,\n",
       " 28706,\n",
       " 29409,\n",
       " 30105,\n",
       " 31325,\n",
       " 31578,\n",
       " 32491,\n",
       " 33170,\n",
       " 33811,\n",
       " 33821,\n",
       " 34031,\n",
       " 34110,\n",
       " 34113,\n",
       " 34299,\n",
       " 34815,\n",
       " 35228,\n",
       " 35246,\n",
       " 35289,\n",
       " 36477,\n",
       " 36609,\n",
       " 38146,\n",
       " 39371,\n",
       " 40348,\n",
       " 40386,\n",
       " 40720,\n",
       " 40795,\n",
       " 42351,\n",
       " 43021,\n",
       " 43067,\n",
       " 43133,\n",
       " 43384,\n",
       " 43987,\n",
       " 44385,\n",
       " 44499,\n",
       " 44548,\n",
       " 44689,\n",
       " 45282,\n",
       " 45628,\n",
       " 45946,\n",
       " 46233,\n",
       " 46683,\n",
       " 46774,\n",
       " 47118,\n",
       " 47124,\n",
       " 47988,\n",
       " 48680,\n",
       " 48917,\n",
       " 49145,\n",
       " 49247,\n",
       " 49679,\n",
       " 50274,\n",
       " 50387,\n",
       " 50577,\n",
       " 51155,\n",
       " 51389,\n",
       " 51513,\n",
       " 51516,\n",
       " 51599,\n",
       " 51698,\n",
       " 52207,\n",
       " 52237,\n",
       " 52348,\n",
       " 53689,\n",
       " 53753,\n",
       " 54437,\n",
       " 54440,\n",
       " 54497,\n",
       " 55001,\n",
       " 55104,\n",
       " 55512,\n",
       " 56220,\n",
       " 57501,\n",
       " 57554,\n",
       " 58232,\n",
       " 58406,\n",
       " 58547,\n",
       " 58661,\n",
       " 58817,\n",
       " 59045,\n",
       " 59677,\n",
       " 59913,\n",
       " 59971,\n",
       " 60210,\n",
       " 60589,\n",
       " 60952,\n",
       " 61204,\n",
       " 62327,\n",
       " 62538,\n",
       " 64399,\n",
       " 66239,\n",
       " 66848,\n",
       " 67955,\n",
       " 68163,\n",
       " 68393,\n",
       " 68736,\n",
       " 69543,\n",
       " 69546,\n",
       " 69558,\n",
       " 70174,\n",
       " 70608,\n",
       " 70716,\n",
       " 70974,\n",
       " 71067,\n",
       " 71325,\n",
       " 71486,\n",
       " 71693,\n",
       " 71831,\n",
       " 71915,\n",
       " 71918,\n",
       " 73419,\n",
       " 74042,\n",
       " 74078,\n",
       " 74119,\n",
       " 74192,\n",
       " 74516,\n",
       " 74537,\n",
       " 74591,\n",
       " 76450,\n",
       " 77378,\n",
       " 77402,\n",
       " 77607,\n",
       " 78747,\n",
       " 78948,\n",
       " 79036,\n",
       " 79681,\n",
       " 80124,\n",
       " 81108,\n",
       " 81354,\n",
       " 81397,\n",
       " 82379,\n",
       " 82481,\n",
       " 83462,\n",
       " 83470,\n",
       " 83492,\n",
       " 83914,\n",
       " 84455,\n",
       " 84478,\n",
       " 84697,\n",
       " 84718,\n",
       " 84841,\n",
       " 84860,\n",
       " 84994,\n",
       " 85091,\n",
       " 85178,\n",
       " 85785,\n",
       " 85956,\n",
       " 86023,\n",
       " 86110,\n",
       " 86485,\n",
       " 86491,\n",
       " 87270,\n",
       " 88239,\n",
       " 88293,\n",
       " 88584,\n",
       " 88862,\n",
       " 90139,\n",
       " 90268,\n",
       " 91372,\n",
       " 91846,\n",
       " 93460,\n",
       " 93766,\n",
       " 94131,\n",
       " 94579,\n",
       " 95792,\n",
       " 96559,\n",
       " 97212,\n",
       " 98410,\n",
       " 99097,\n",
       " 99350,\n",
       " 99604,\n",
       " 100042,\n",
       " 100933,\n",
       " 101404,\n",
       " 101513,\n",
       " 102761,\n",
       " 103151,\n",
       " 103193,\n",
       " 103259,\n",
       " 103268,\n",
       " 103439,\n",
       " 103565,\n",
       " 103634,\n",
       " 104119,\n",
       " 104700,\n",
       " 104985,\n",
       " 105739,\n",
       " 105772,\n",
       " 106025,\n",
       " 106152,\n",
       " 106173,\n",
       " 106281,\n",
       " 106382,\n",
       " 106535,\n",
       " 107013,\n",
       " 107241,\n",
       " 107801,\n",
       " 109154,\n",
       " 109397,\n",
       " 109661,\n",
       " 109682,\n",
       " 110591,\n",
       " 110692,\n",
       " 111436,\n",
       " 111502,\n",
       " 112006,\n",
       " 113071,\n",
       " 113165,\n",
       " 113219,\n",
       " 113312,\n",
       " 113847,\n",
       " 113961,\n",
       " 114312,\n",
       " 114855,\n",
       " 116019,\n",
       " 116556,\n",
       " 118041,\n",
       " 118581,\n",
       " 118638,\n",
       " 120477,\n",
       " 120480,\n",
       " 121946,\n",
       " 122176,\n",
       " 122645,\n",
       " 122855,\n",
       " 123742,\n",
       " 123945,\n",
       " 124515,\n",
       " 124521,\n",
       " 124524,\n",
       " 125458,\n",
       " 129423,\n",
       " 129954,\n",
       " 131342,\n",
       " 131525,\n",
       " 133012,\n",
       " 133101,\n",
       " 133941,\n",
       " 134251,\n",
       " 134269,\n",
       " 135229,\n",
       " 136829,\n",
       " 138945,\n",
       " 139204,\n",
       " 140185,\n",
       " 140583,\n",
       " 141424,\n",
       " 141549,\n",
       " 141781,\n",
       " 141912,\n",
       " 142039,\n",
       " 142093,\n",
       " 142369,\n",
       " 142560,\n",
       " 142603,\n",
       " 142944,\n",
       " 143424,\n",
       " 143470,\n",
       " 143919,\n",
       " 144785,\n",
       " 145088,\n",
       " 145382,\n",
       " 145452,\n",
       " 145757,\n",
       " 146671,\n",
       " 147148,\n",
       " 147510,\n",
       " 148146,\n",
       " 148542,\n",
       " 148910,\n",
       " 149274,\n",
       " 149355,\n",
       " 149857,\n",
       " 150725,\n",
       " 150860,\n",
       " 151025,\n",
       " 151109,\n",
       " 151557,\n",
       " 151703,\n",
       " 151799,\n",
       " 151822,\n",
       " 151845,\n",
       " 151867,\n",
       " 151890,\n",
       " 151913,\n",
       " 151999,\n",
       " 152013,\n",
       " 152134,\n",
       " 152355,\n",
       " 152394,\n",
       " 152986,\n",
       " 153147,\n",
       " 153570,\n",
       " 153832,\n",
       " 154153,\n",
       " 154214,\n",
       " 154511,\n",
       " 154582,\n",
       " 154793,\n",
       " 155341,\n",
       " 155572,\n",
       " 155817,\n",
       " 156177,\n",
       " 157487,\n",
       " 157526,\n",
       " 157553,\n",
       " 158104,\n",
       " 158632,\n",
       " 160238,\n",
       " 160376,\n",
       " 161006,\n",
       " 161024,\n",
       " 161240,\n",
       " 161300,\n",
       " 161668,\n",
       " 162365,\n",
       " 162853,\n",
       " 163969,\n",
       " 164023,\n",
       " 164251,\n",
       " 166151,\n",
       " 166204,\n",
       " 166463,\n",
       " 166504,\n",
       " 166726,\n",
       " 167040,\n",
       " 167082,\n",
       " 167960,\n",
       " 168024,\n",
       " 168482,\n",
       " 168676,\n",
       " 169595,\n",
       " 169689,\n",
       " 170096,\n",
       " 170333,\n",
       " 170829,\n",
       " 170937,\n",
       " 171186,\n",
       " 171306,\n",
       " 171942,\n",
       " 173200,\n",
       " 173383,\n",
       " 173976,\n",
       " 173995,\n",
       " 174132,\n",
       " 174322,\n",
       " 174597,\n",
       " 174790,\n",
       " 175291,\n",
       " 175325,\n",
       " 175589,\n",
       " 175956,\n",
       " 175964,\n",
       " 176181,\n",
       " 176504,\n",
       " 177264,\n",
       " 177542,\n",
       " 177600,\n",
       " 177830,\n",
       " 178010,\n",
       " 178367,\n",
       " 178489,\n",
       " 178767,\n",
       " 179217,\n",
       " 179413,\n",
       " 179629,\n",
       " 179909,\n",
       " 180167,\n",
       " 180197,\n",
       " 180367,\n",
       " 180652,\n",
       " 181322,\n",
       " 181643,\n",
       " 181730,\n",
       " 181873,\n",
       " 182269,\n",
       " 182842,\n",
       " 183334,\n",
       " 183806,\n",
       " 183964,\n",
       " 184150,\n",
       " 184411,\n",
       " 184759,\n",
       " 184819,\n",
       " 185134,\n",
       " 185140,\n",
       " 185215,\n",
       " 185221,\n",
       " 185292,\n",
       " 185383,\n",
       " 185394,\n",
       " 185549,\n",
       " 186146,\n",
       " 186189,\n",
       " 186522,\n",
       " 186561,\n",
       " 187930,\n",
       " 188281,\n",
       " 188533,\n",
       " 188623,\n",
       " 189422,\n",
       " 189556,\n",
       " 190460,\n",
       " 191270,\n",
       " 191424,\n",
       " 191616,\n",
       " 192109,\n",
       " 192967,\n",
       " 193629,\n",
       " 194276,\n",
       " 194676,\n",
       " 195227,\n",
       " 195233,\n",
       " 195469,\n",
       " 196466,\n",
       " 197293,\n",
       " 197878,\n",
       " 198066,\n",
       " 198897,\n",
       " 199176,\n",
       " 200076,\n",
       " 200161,\n",
       " 200338,\n",
       " 201025,\n",
       " 201325,\n",
       " 201355,\n",
       " 201847,\n",
       " 202002,\n",
       " 202087,\n",
       " 202860,\n",
       " 204124,\n",
       " 204838,\n",
       " 205525,\n",
       " 205849,\n",
       " 206841,\n",
       " 206942,\n",
       " 207200,\n",
       " 208443,\n",
       " 209149,\n",
       " 209248,\n",
       " 209524,\n",
       " 209540,\n",
       " 210008,\n",
       " 210688,\n",
       " 211048,\n",
       " 213801,\n",
       " 214440,\n",
       " 216454,\n",
       " 217113,\n",
       " 217416,\n",
       " 217983,\n",
       " 218301,\n",
       " 218408,\n",
       " 218700,\n",
       " 219380,\n",
       " 219401,\n",
       " 223350,\n",
       " 224187,\n",
       " 224486,\n",
       " 224527,\n",
       " 224987,\n",
       " 225050,\n",
       " 225221,\n",
       " 225620,\n",
       " 226564,\n",
       " 226586,\n",
       " 227888,\n",
       " 228243,\n",
       " 228884,\n",
       " 229098,\n",
       " 229872,\n",
       " 231953,\n",
       " 232698,\n",
       " 233192,\n",
       " 234052,\n",
       " 234873,\n",
       " 235291,\n",
       " 235676,\n",
       " 236439,\n",
       " 236665,\n",
       " 237404,\n",
       " 237603,\n",
       " 237779,\n",
       " 237899,\n",
       " 238202,\n",
       " 238247,\n",
       " 238785,\n",
       " 239321,\n",
       " 239438,\n",
       " 239599,\n",
       " 240703,\n",
       " 240854,\n",
       " 241053,\n",
       " 241138,\n",
       " 241669,\n",
       " 242091,\n",
       " 242293,\n",
       " 243484,\n",
       " 244777,\n",
       " 245124,\n",
       " 245769,\n",
       " 246873,\n",
       " 247515,\n",
       " 248054,\n",
       " 248430,\n",
       " 249039,\n",
       " 249267,\n",
       " 249520,\n",
       " 249606,\n",
       " 249701,\n",
       " 249849,\n",
       " 249855,\n",
       " 250793,\n",
       " 252213,\n",
       " 252975,\n",
       " 253174,\n",
       " 253563,\n",
       " 254454,\n",
       " 255806,\n",
       " 256831,\n",
       " 256924,\n",
       " 259734,\n",
       " 259927,\n",
       " 259963,\n",
       " 260264,\n",
       " 260486,\n",
       " 260546,\n",
       " 260645,\n",
       " 260846,\n",
       " 261197,\n",
       " 261585,\n",
       " 261945,\n",
       " 262025,\n",
       " 262185,\n",
       " 262199,\n",
       " 262480,\n",
       " 262582,\n",
       " 262739,\n",
       " 263007,\n",
       " 263128,\n",
       " 263380,\n",
       " 263759,\n",
       " 263967,\n",
       " 264027,\n",
       " 264193,\n",
       " 264280,\n",
       " 264520,\n",
       " 264700,\n",
       " 266690,\n",
       " 266951,\n",
       " 267210,\n",
       " 267408,\n",
       " 267476,\n",
       " 267617,\n",
       " 268142,\n",
       " 268203,\n",
       " 268897,\n",
       " 271209,\n",
       " 271440,\n",
       " 271486,\n",
       " 271496,\n",
       " 271729,\n",
       " 271980,\n",
       " 273671,\n",
       " 274253,\n",
       " 274291,\n",
       " 274373,\n",
       " 274430,\n",
       " 274829,\n",
       " 275264,\n",
       " 275384,\n",
       " 275928,\n",
       " 277792,\n",
       " 278520,\n",
       " 278548,\n",
       " 278932,\n",
       " 279372,\n",
       " 279757,\n",
       " 281752,\n",
       " 281954,\n",
       " 282275,\n",
       " 282296,\n",
       " 282664,\n",
       " 283564,\n",
       " 284230,\n",
       " 284793,\n",
       " 284905,\n",
       " 285303,\n",
       " 285388,\n",
       " 285789,\n",
       " 286015,\n",
       " 287255,\n",
       " 287514,\n",
       " 288397,\n",
       " 289331,\n",
       " 290141,\n",
       " 291381,\n",
       " 292122,\n",
       " 292172,\n",
       " 293623,\n",
       " 294179,\n",
       " 295069,\n",
       " 295685,\n",
       " 295807,\n",
       " 296388,\n",
       " 296970,\n",
       " 297083,\n",
       " 297201,\n",
       " 297648,\n",
       " 297862,\n",
       " 299285,\n",
       " 299413,\n",
       " 299737,\n",
       " 299870,\n",
       " 300498,\n",
       " 301237,\n",
       " 301475,\n",
       " 301589,\n",
       " 301721,\n",
       " 301733,\n",
       " 301814,\n",
       " 302171,\n",
       " 302326,\n",
       " 302914,\n",
       " 303107,\n",
       " 303203,\n",
       " 303420,\n",
       " 303621,\n",
       " 303986,\n",
       " 304088,\n",
       " 304404,\n",
       " 304950,\n",
       " 305166,\n",
       " 305195,\n",
       " 305593,\n",
       " 305863,\n",
       " 306163,\n",
       " 306564,\n",
       " 306788,\n",
       " 306942,\n",
       " 307295,\n",
       " 307692,\n",
       " 308992,\n",
       " 309721,\n",
       " 309796,\n",
       " 309916,\n",
       " 310027,\n",
       " 310535,\n",
       " 311391,\n",
       " 311528,\n",
       " 312769,\n",
       " 314175,\n",
       " 315046,\n",
       " 315449,\n",
       " 316032,\n",
       " 316338,\n",
       " 316481,\n",
       " 317435,\n",
       " 317931,\n",
       " 317942,\n",
       " 318261,\n",
       " 318495,\n",
       " 318805,\n",
       " 318922,\n",
       " 318966,\n",
       " 318979,\n",
       " 319085,\n",
       " 319543,\n",
       " 319577,\n",
       " 320510,\n",
       " 320855,\n",
       " 321635,\n",
       " 321748,\n",
       " 322031,\n",
       " 322118,\n",
       " 323282,\n",
       " 323426,\n",
       " 323809,\n",
       " 324516,\n",
       " 324968,\n",
       " 325515,\n",
       " 326025,\n",
       " 326308,\n",
       " 326500,\n",
       " 326821,\n",
       " 327180,\n",
       " 327835,\n",
       " 327947,\n",
       " 328661,\n",
       " 328948,\n",
       " 329698,\n",
       " 330390,\n",
       " 332983,\n",
       " 333330,\n",
       " 333922,\n",
       " 334064,\n",
       " 334195,\n",
       " 334543,\n",
       " 335727,\n",
       " 336612,\n",
       " 336638,\n",
       " 336722,\n",
       " 337371,\n",
       " 337482,\n",
       " 338684,\n",
       " 338687,\n",
       " 338806,\n",
       " 339034,\n",
       " 339706,\n",
       " 339717,\n",
       " 339817,\n",
       " 340564,\n",
       " 340609,\n",
       " 341068,\n",
       " 342139,\n",
       " 342927,\n",
       " 343092,\n",
       " 343339,\n",
       " 343615,\n",
       " 343862,\n",
       " 344366,\n",
       " 344648,\n",
       " 344753,\n",
       " 345014,\n",
       " 345188,\n",
       " 345745,\n",
       " 346148,\n",
       " 346470,\n",
       " 346746,\n",
       " 346997,\n",
       " 348078,\n",
       " 348339,\n",
       " 350312,\n",
       " 350330,\n",
       " 351556,\n",
       " 351595,\n",
       " 352020,\n",
       " 352123,\n",
       " 352279,\n",
       " 352303,\n",
       " 352354,\n",
       " 352498,\n",
       " 353163,\n",
       " 353888,\n",
       " 355031,\n",
       " 355041,\n",
       " 355154,\n",
       " 355493,\n",
       " 355613,\n",
       " 355855,\n",
       " 356179,\n",
       " 357539,\n",
       " 357905,\n",
       " 358083,\n",
       " 358300,\n",
       " 358775,\n",
       " 359418,\n",
       " 359468,\n",
       " 359577,\n",
       " 359803,\n",
       " 360265,\n",
       " 360456,\n",
       " 360736,\n",
       " 361259,\n",
       " 361596,\n",
       " 361692,\n",
       " 361734,\n",
       " 362242,\n",
       " 362443,\n",
       " 362791,\n",
       " 363007,\n",
       " 363419,\n",
       " 363737,\n",
       " 363767,\n",
       " 364161,\n",
       " 364635,\n",
       " 364642,\n",
       " 365437,\n",
       " 365867,\n",
       " 366372,\n",
       " 367338,\n",
       " 367626,\n",
       " 367752,\n",
       " 367809,\n",
       " 367959,\n",
       " 368546,\n",
       " 368691,\n",
       " 368759,\n",
       " 368927,\n",
       " 370801,\n",
       " 371369,\n",
       " 372237,\n",
       " 372366,\n",
       " 372603,\n",
       " 373649,\n",
       " 373761,\n",
       " 374439,\n",
       " 375196,\n",
       " 375229,\n",
       " 375339,\n",
       " 375412,\n",
       " 375631,\n",
       " 376607,\n",
       " 377381,\n",
       " 378054,\n",
       " 378467,\n",
       " 378983,\n",
       " 379344,\n",
       " 380322,\n",
       " 380331,\n",
       " 381006,\n",
       " 381435,\n",
       " 382027,\n",
       " 382836,\n",
       " 383767,\n",
       " 384412,\n",
       " 386284,\n",
       " 386621,\n",
       " 387874,\n",
       " 389273,\n",
       " 389430,\n",
       " 389590,\n",
       " 389788,\n",
       " 390790,\n",
       " 391653,\n",
       " 392196,\n",
       " 392596,\n",
       " 393034,\n",
       " 393810,\n",
       " 395858,\n",
       " 395864,\n",
       " 396000,\n",
       " 396567,\n",
       " 396636,\n",
       " 396831,\n",
       " 397430,\n",
       " 398181,\n",
       " 398440,\n",
       " 398461,\n",
       " 398525,\n",
       " 398621,\n",
       " 398741,\n",
       " 398912,\n",
       " 399449,\n",
       " 399737,\n",
       " 399956,\n",
       " 399987,\n",
       " 400213,\n",
       " 400769,\n",
       " 401416,\n",
       " 401448,\n",
       " 401625,\n",
       " 401716,\n",
       " 402394,\n",
       " 403846,\n",
       " 403875,\n",
       " 404252,\n",
       " 405359,\n",
       " 405482,\n",
       " 406007,\n",
       " 406035,\n",
       " 406221,\n",
       " 407331,\n",
       " 408868,\n",
       " 409338,\n",
       " 409684,\n",
       " 409981,\n",
       " 410075,\n",
       " 412257,\n",
       " 412744,\n",
       " 413605,\n",
       " 413611,\n",
       " 414277,\n",
       " 415087,\n",
       " 415670,\n",
       " 415826,\n",
       " 416300,\n",
       " 416567,\n",
       " 416784,\n",
       " 417472,\n",
       " 417485,\n",
       " 417947,\n",
       " 418019,\n",
       " 418225,\n",
       " 418727,\n",
       " 419103,\n",
       " 419304,\n",
       " 420364,\n",
       " 420406,\n",
       " 420577,\n",
       " 420833,\n",
       " 420866,\n",
       " 420875,\n",
       " 420977,\n",
       " 421277,\n",
       " 421667,\n",
       " 421739,\n",
       " 423211,\n",
       " 423214,\n",
       " 423245,\n",
       " 423329,\n",
       " 424171,\n",
       " 424603,\n",
       " 424768,\n",
       " 425215,\n",
       " 425359,\n",
       " 425548,\n",
       " 425743,\n",
       " 426394,\n",
       " 426700,\n",
       " 427036,\n",
       " 427084,\n",
       " 427132,\n",
       " 427270,\n",
       " 427495,\n",
       " 427858,\n",
       " 427945,\n",
       " 427999,\n",
       " 428065,\n",
       " 429508,\n",
       " 429622,\n",
       " 429781,\n",
       " ...]"
      ]
     },
     "metadata": {},
     "execution_count": 7
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "source": [
    "bzy.approx_pattern_count(\"atgatcaag\",vc_genome_short,2)"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "2581"
      ]
     },
     "metadata": {},
     "execution_count": 8
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "## Caveat: For these scripts, All 4k k-mers Pattern, compute approx_pattern_count for each k-mer Pattern, and then find k-mers with the maximum number of approximate occurrences. This is an inefficient approach, since many of the 4k k-mers should not be considered because neither they nor their mutated versions (with up to d mismatches) appear in Text.\n",
    "\n",
    "## Instead, the algorithm freq_words_mismatch will be used. It uses a single map that counts the number of times a given string has an approximate match in Text. For a given k-mer substring Pattern of Text, we need to increase 1 to the count of every k-mer that has Hamming distance at most d from Pattern. Note here, the vc_ori is used as oppose to Ecoli_genome.txt. My algorithm's runtime is too slow to work with the million length dataset.\n"
   ],
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "source": [
    "bzy.freq_words_mismatch(vc_ori,9,1)"
   ],
   "outputs": [
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "['Catgatcaa',\n",
       " 'Gatgatcaa',\n",
       " 'Aatgatcaa',\n",
       " 'Tatgatcaa',\n",
       " 'atgatcaaG',\n",
       " 'atgatcaaA',\n",
       " 'atgatcaaT',\n",
       " 'atgatcaaC',\n",
       " 'atgatcaGg',\n",
       " 'atgatcaTg',\n",
       " 'atgatcaCg',\n",
       " 'atgatcaAg',\n",
       " 'cGtgatcat',\n",
       " 'cCtgatcat',\n",
       " 'cTtgatcat',\n",
       " 'cAtgatcat']"
      ]
     },
     "metadata": {},
     "execution_count": 11
    }
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Accounting for reverse compliments as well, it is found that within a window of length 500 starting at position 3923620 (3923620 - 3924120) of the E. coli genome, the experimentally confirmed DnaA box in E. coli (TTATCCACA) is indeed a most frequent 9-mer with 1 mismatch, along with its reverse complement TGTGGATAA\n",
    "\n",
    "## Possible reason for finding multiple other frequent 9mers when looking for ori: \n",
    "\n",
    "1. The other 9-mers are for binding other proteins, e.g. a transcription enzyme.\n",
    "\n",
    "2. DNA Polymerase looks for more than just one DnaA box, and it will recognize different sequences in different boxes.\n",
    "\n",
    "3. The actual DnaA box is less than 9 bases.  I see strings of C's or G's in the middle, so maybe the surrounding bases don't matter as much.\n",
    "\n",
    "4. There are more than one DNA Polymerase in e. coli, and they each have different DnaA boxes, and there is some kind of selective pressure to make sure that they all bind in the same region."
   ],
   "metadata": {}
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Complications in locating the ori\n",
    "\n",
    "1.  Some bacteria have even fewer DnaA boxes than E. coli, making it difficult to identify them.\n",
    "\n",
    "2. The ter region is often located not directly opposite to ori but may be significantly shifted, resulting in reverse and forward half-strands having substantially different lengths. \n",
    "\n",
    "3.  The position of the skew minimum is often only a rough indicator of ori position, which forces researchers to expand their windows when searching for DnaA boxes, bringing in extraneous repeated substrings.\n",
    "\n",
    "4. Skew Diagrams often converge to local minimums due to different evolutionary factors. For example, the skew diagram for Thermotoga petrophila fluctuates heavily. I deduce that as an extremophile, it is adapted to very hot living conditions. As a result, its DNA has to be more stable and hence have a higher GC percentage."
   ],
   "metadata": {}
  }
 ],
 "metadata": {
  "orig_nbformat": 4,
  "language_info": {
   "name": "python",
   "version": "3.9.5",
   "mimetype": "text/x-python",
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "pygments_lexer": "ipython3",
   "nbconvert_exporter": "python",
   "file_extension": ".py"
  },
  "kernelspec": {
   "name": "python3",
   "display_name": "Python 3.9.5 64-bit ('venv-bioezy-test': virtualenvwrapper)"
  },
  "interpreter": {
   "hash": "044f29bb47c2120b5fe22aba225e4b72523c7088c7615f65b1801c6ace85334e"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}